### ISSUE TITLE
5.7 toolchain from 5. July reports warning if non-sendable local function is passed to a `@preconcurrency` marked method

### ISSUE THREAD

**Issue by dnadoba** [apple/swift#59910]
**Describe the bug**
The Swift 5.7 toolchain from 5. July for Xcode and Ubuntu produce a new warning:

```
/src/Sources/NIOHTTP1Server/main.swift:234:67: warning: converting non-sendable function value to '@Sendable () throws -> ()' may introduce data races
                        context.eventLoop.scheduleTask(in: delay, doNext)
```
Code can be found here: https://github.com/apple/swift-nio/blob/f1d62128c4587a06bee9f5455eef282694b71288/Sources/NIOHTTP1Server/main.swift#L234

[`EventLoop.scheduleTask(in:_)` is marked with `@preconcurrency`](https://github.com/apple/swift-nio/blob/f1d62128c4587a06bee9f5455eef282694b71288/Sources/NIOCore/EventLoop.swift#L327-L328) and therefore (AFAIK) should not produce a warning in 5.7 without `-warn-concurrency`.

**Steps To Reproduce**
- Clone swift-nio from https://github.com/apple/swift-nio.git
- open Package.swift with Xcode and build the package (or just `NIOHTTP1Server`)

Or with docker:
```
git clone https://github.com/apple/swift-nio
cd swift-nio
docker run --rm --volume "$(pwd)/:/src" --workdir "/src/" swiftlang/swift:nightly-5.7-focal swift build
```

**Expected behavior**
No warning should be emitted

**Screenshots**
<img width="1369" alt="Screen Shot 2022-07-06 at 14 31 43" src="https://user-images.githubusercontent.com/5688874/177550652-46fabcb9-70eb-432a-82c0-be4e67853dba.png">

**Environment (please fill out the following information)**
 - OS: 13.0 Beta (22A5266r)
 - Xcode Version 14.0 beta 3 (14A5270f) with Swift 5.7 Development Snapshot 2022-07-05 (a)

or Docker with swiftlang/swift:nightly-5.7-focal on arm64

**Additional context**
Didn't produce these warnings in previous 5.7 snapshots

Other issues new in the 5.7 2022-07-05 snapshot:

https://github.com/apple/swift/issues/59909
https://github.com/apple/swift/issues/59911
https://github.com/apple/swift/issues/59913

**Comment by DougGregor** [comment]
The actual issue here is that the closure to `whenSuccess` is known to be `@Sendable` based on context, but the solver itself isn't tracking that the `@Sendable` comes from a `@preconcurrency` call. This is likely to be the same problem as https://github.com/apple/swift/issues/59911.

**Comment by DougGregor** [comment]
Reduction:

```swift
struct Scheduled<T> { }

@preconcurrency
func doPreconcurrency(_: @Sendable () -> Void) { }

class EventLoop {
  @discardableResult
  @preconcurrency
  func scheduleTask<T>(deadline: Int, _ task: @escaping @Sendable () throws -> T) -> Scheduled<T> { fatalError("") }
}

class C {
  var ev: EventLoop? = nil

  func test(i: Int) {
    func doNext() {
      doPreconcurrency {
        self.ev?.scheduleTask(deadline: i, doNext)
        return
      }
    }
  }
}
```

**Comment by dnadoba** [comment]
Fixed in the latest 5.7 toolchain from July 17, 2022
